### PR TITLE
Bitfinex Brokerage updates

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
- Upgrade Bitfinex API to v2 (current stable and recommended version)
- Order submit/cancel/update using WebSocket instead of REST (to prevent delays caused by rate limits)
- Removed URL config settings for WebSocket and REST API endpoints (for easier updating in the future)
- Bugfix: no longer emits trade ticks on trades snapshot message

#### Related Issue
n/a

#### Motivation and Context
- Live trading stability

#### Requires Documentation Change
No.

#### How Has This Been Tested?
- Brokerage unit tests (all green)
- Live trading locally + QC cloud
- Tested streaming for all symbols in QC backend application

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] All new and existing tests passed.

### CODE REVIEW

File: Brokerages/Bitfinex/BitfinexSubscriptionManager.cs [hunk 1]
```diff
@@ -319,26 +319,45 @@ private void OnMessage(object sender, WebSocketMessage e)
...
+                            case "hb":
+                                return;
+
+                            // trade
+                            case "tu":
+                                OnUpdate(webSocket, channel.ToStringInvariant(), token[2].ToObject<string[]>());
+                                break;
+
+                            // ignore
+                            case "te":
```
Comment (Martin-Molinero) [inline Brokerages/Bitfinex/BitfinexSubscriptionManager.cs:344]: Wondering what's `te`? Maybe we should add why we ignore it in the comment

File: Brokerages/Bitfinex/BitfinexSubscriptionManager.cs [hunk 2]
```diff
@@ -462,9 +485,6 @@ private void OnSnapshot(BitfinexWebSocketWrapper webSocket, string channelId, st
                     case "book":
                         ProcessOrderBookSnapshot(channel, entries);
                         return;
-                    case "trades":
-                        ProcessTradesSnapshot(channel, entries);
```
Comment (Martin-Molinero) [inline Brokerages/Bitfinex/BitfinexSubscriptionManager.cs:466]: Wondering why was this a bug?

File: Brokerages/Bitfinex/BitfinexBrokerage.cs [hunk 3]
```diff
@@ -175,10 +201,10 @@ public override List<Order> GetOpenOrders()
                     continue;
                 }
 
-                order.Quantity = item.Side == "sell" ? -item.OriginalAmount : item.OriginalAmount;
-                order.BrokerId = new List<string> { item.Id };
+                order.Quantity = item.Amount;
+                order.BrokerId = new List<string> { item.Id.ToStringInvariant() };
                 order.Symbol = _symbolMapper.GetLeanSymbol(item.Symbol);
-                order.Time = Time.UnixTimeStampToDateTime(item.Timestamp);
+                order.Time = Time.UnixMillisecondTimeStampToDateTime(item.MtsCreate);
```
Comment (Martin-Molinero) [inline Brokerages/Bitfinex/BitfinexBrokerage.cs:207]: We might be able to set the LastUpdateTime with `MtsUpdate` I think, maybe if different from `MtsCreate`?

File: Brokerages/QuantConnect.Brokerages.csproj [hunk 4]
```diff
@@ -352,8 +352,19 @@
...
+    <Compile Include="Bitfinex\Messages\Position.cs" />
+    <Compile Include="Bitfinex\Messages\Ticker.cs" />
+    <Compile Include="Bitfinex\Messages\TradeExecutionUpdate.cs" />
+    <Compile Include="Bitfinex\Messages\Wallet.cs" />
+    <Compile Include="Bitfinex\Converters\TradeExecutionUpdateConverter.cs" />
+    <Compile Include="Bitfinex\Converters\OrderConverter.cs" />
+    <Compile Include="Bitfinex\Messages\OrderFlags.cs" />
+    <Compile Include="Bitfinex\Converters\TickerConverter.cs" />
+    <Compile Include="Bitfinex\Converters\PositionConverter.cs" />
+    <Compile Include="Bitfinex\Converters\WalletConverter.cs" />
```
Comment (Martin-Molinero) [inline Brokerages/QuantConnect.Brokerages.csproj:367]: Love the new converters 👍

File: Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs [hunk 5]
```diff
@@ -37,16 +39,21 @@ namespace QuantConnect.Brokerages.Bitfinex
...
-        private volatile bool _streamLocked;
-        private readonly RateGate _restRateLimiter = new RateGate(8, TimeSpan.FromMinutes(1));
+        private readonly RateGate _restRateLimiter = new RateGate(10, TimeSpan.FromMinutes(1));
         private readonly ConcurrentDictionary<int, decimal> _fills = new ConcurrentDictionary<int, decimal>();
         private readonly BitfinexSubscriptionManager _subscriptionManager;
         private readonly SymbolPropertiesDatabase _symbolPropertiesDatabase;
         private readonly IDataAggregator _aggregator;
 
+        // map ClientOrderId -> LEAN order
+        private readonly ConcurrentDictionary<long, Order> _orderMap = new ConcurrentDictionary<long, Order>();
```
Comment (Martin-Molinero) [inline Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs:54]: Seems a bit of duplication with `CachedOrderIDs`? We are not removing any order from this collection.
On `GetOpenOrders` I only see we update `CachedOrderIDs` if already present? Should we be adding them instead 🤔 same for `_orderMap`

**Review by Martin-Molinero — COMMENTED** [review]
Looks good! Leaving some concerns and questions. Sorry for the late review 😅

File: Brokerages/Bitfinex/BitfinexSubscriptionManager.cs (hunk 1 above)
Comment (StefanoRaggi) [reply Brokerages/Bitfinex/BitfinexSubscriptionManager.cs:344]: The `te` message is a trade execution, similar to `tu` (trade execution update):
both messages are sent for each order fill but only the `tu` message contains fields for the fees.
I'll add a comment to clarify

File: Brokerages/Bitfinex/BitfinexSubscriptionManager.cs (hunk 2 above)
Comment (StefanoRaggi) [reply Brokerages/Bitfinex/BitfinexSubscriptionManager.cs:466]: When successfully subscribing to a symbol, the exchange sends a snapshot containing the latest 20-40 ticks (can't remember the exact number), this is historical data which should not be emitted as live ticks.

File: Brokerages/Bitfinex/BitfinexBrokerage.cs (hunk 3 above)
Comment (StefanoRaggi) [reply Brokerages/Bitfinex/BitfinexBrokerage.cs:207]: I'll check if `MtsUpdate` is also updated on each fill, in this case we should update `LastFillTime` instead

File: Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs (hunk 5 above)
Comment (StefanoRaggi) [reply Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs:54]: In `_orderMap` the key is not the Lean `OrderId` but the Bitfinex `ClientOrderId` field (`cid`) which is used to match async messages. 
In the docs they mention this field is meant to be unique in the (although not enforced):
https://docs.bitfinex.com/reference#ws-auth-input-order-new

I think using the `CachedOrderIDs` dictionary, we could have duplicate id issues with existing orders, if an algorithm is started twice on the same day, maybe should use a timestamp instead?

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**